site-src/geps/gep-1282.md

```diff
@@ -0,0 +1,79 @@
+
+
+## Non-Goals
+
+* We don’t want to require changes to v1.Service to start with, as the turnaround time for those changes can be years.
+* Our solution should not use unstructured data like labels and annotations - these are very sticky and hard to get rid of once you start using them.
+* v1.Service’s appProtocol field is not fit for purpose, because it is defined as accepting values either from the [IANA Service Name registry](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtmly), or domain-prefixed values and we need more flexibility than that.
+* To be populated as we discuss more.
+
+
+## Introduction
+
+As implementations have started to build out HTTPRoute support, requests for some common capabilities have started coming up, particularly TLS re-encryption (encryption between Gateway and backend), Websockets tracking, and HTTP/2 support.
+
+Evan Anderson opened a [discussion that generated a lot of interest](https://github.com/kubernetes-sigs/gateway-api/discussions/1244), and during that discussion, we brought up a few more points:
+* Whatever we do to solve this in the HTTPRoute case has applicability to describing more general service properties, like identity, which could be useful for other Route use cases.
+* This may be very useful for mesh/GAMMA use cases as well as more generally for adding arbitrary future capabilities to the backend service. An example is a CA cert for connecting to the backend - that’s tightly bound to the service, but there’s nowhere to put it at the moment.
+
+This initial form of this GEP is for the Gateway API community to agree on what problem we’re solving and why.
+
+### What we're building
+
+We’re looking to add specific, structured extension points somewhere in the resources used by the Gateway API to describe properties and capabilities of backend services.
+
+The initial list includes, but is not limited to:
+* TLS information for connection from the Gateway to the backend service. Note that this doesn’t include any information used for service mesh encryption, just what a Gateway’s proxy would need to be able to connect to the backend service.
+* Websocket protocol information for the backend service.
+* Protocol disambiguation for “upgradeable” protocols like HTTP/2 and HTTP/1.1 which operate on the same port. (Websockets may be another case of this.)
+
+### Why build something?
+
+The existing resources we have (Service and HTTPRoute) don’t have space for this information currently, and we have feedback from users via various implementations that it’s necessary - you could argue that some of this is table stakes for a layer 7 ingress controller.
+
+In previous discussions, we discussed a variety of ways to achieve this goal, but there’s been a lot of discussion already, so we’re moving to a GEP discussion format to try to capture feedback more efficiently.
```

Comment:
Right now, I think this "why" section needs significantly more content, it doesn't seem to be fully supporting the "why should we do this" question:

> The existing resources we have (Service and HTTPRoute) don’t have space for this information currently

This just extends the "what": we're reiterating _what_ we want to see done by stating that it doesn't exist yet, but that isn't part of an answer to "why".

> and we have feedback from users via various implementations that it’s necessary

This comment needs some quantification, can we add the user feedback we've received here (even if paraphrased some) into some specific needs that justify the extra service metadata we're proposing?

> you could argue that some of this is table stakes for a layer 7 ingress controller.

This comment just seems to reiterate the desire of the author, but we need to quantify this: what capabilities of an L7 ingress controller demand the service metadata we're proposing?

>  In previous discussions, we discussed a variety of ways to achieve this goal, but there’s been a lot of discussion already, so we’re moving to a GEP discussion format to try to capture feedback more efficiently.

This doesn't seem to belong in "why": it seems to be more of a note to indicate there have been previous discussions (without links)?

Reply:
It seems like what we need here is:
- a list of the actual feature requests we have in this repo
- examples of how other ingress controllers have requests for similar things
- More links to evidence.

Does that sound right?

Reply:
Sure, also potentially user stories. Ideally some stuff more written from the perspective of the cluster operator who has to set the ingress up for other teams.